torch/library.py

```diff
@@ -473,6 +473,25 @@ def inner(func):
         return inner
     return inner(func)
 
+registered_class: Dict[str, Any] = {}
+
+def impl_abstract_class(qualname, fake_class=None, *, lib=None, _stacklevel=1):
```

Comment:
will refactor this following impl_atract

Reply:
We might want to think about what the best API for this is. Our options are:
- impl_abstract_class, like impl_abstract
- a method directly on the torchbind object.

Also, we really want two things:
1. a way to get the "state/metadata" of a TorchBind class
2. a way to convert the "state/metadata" into a Fake version of the TorchBind class.

I don't see those two things here (we want (1) by itself so that we can use it for guarding)

Reply:
> a method directly on the torchbind object.

Are your referring to something like:
```python
class FakeFoo
  ...
foo_obj = torch.classes._Testing.Foo()
foo_obj.impl_fake(FakeFoo) # Or foo_obj.impl_fake(FakeFoo.from_real(foo_obj))?  
```
What's bad for `impl_abstrct_class`? It seems consistent with existing APIs. Does second approach bring additional benefits? My understanding is that it creates a direct relation between the TorchScript object and its corresponding fake implementation and perhaps less verbose compared with `impl_abstract_class`.

> a way to get state and convert to fake with meta data.

I'll add them shortly.

> create in a private namespace

Sounds good!

Reply:
> I'll add them shortly. After we have the get_metadata and fake_from_metadata, we no longer need from_real, right?

Yup 

> Are your referring to something like:

Something like:
```py
foo_obj = torch.classes._Testing.Foo()
metadata = foo_obj.__get_metadata__() 
fake_foo = foo_obj.__fake_from_metadata__(metadata)
```

> What's bad for impl_abstrct_class? It seems consistent with existing APIs. Does second approach bring additional benefits?

For tensor subclasses we defined `__tensor_flatten__`/`__tensor_unflatten__` as the way to retrieve and work with internal state. For operators we have impl_abstract. It's a question of which approach is "more consistent"

Reply:
The interface looks good to me! Yeah, the consistency with the tensors are indeed better analogy than classes. Let me get it working.

Reply:
It's unclear to me if `__get_metadata__/__fake_from_metadata__` are better, because torchbind objects get defined in C++...

Reply:
Yeah. To summarize: 
1. We need `__get_metadata__` anyway in dynamo guards so users have to implement this method. 
2. Users would also have to write a fake class in Python. 
3. When we encounter a script_object args for a custom op, we need a way to map the script_object to its fake class. The most straightforward way would be using the impl_atract_class apporach. Therefore users need to register this mapping.
4. users need to tell us how to create the fake class. Probably, we could ask users to implement fake_class.__fake_from_metadata__, where users must access the metadata via `__get_metadate__`. 

The end user-journey looks like this:
1. User pt2 compiles a model with torchScriptObject and sees an error message like ` cannot fakify the scriptObject and please register a fake class that implements a static method from_metadata`.
2. To write this fake class, users go back to C++ implementation and tries to collect the metadata required and add a __get_metadate__ methods with def_get_metadata. 
3. Then user implements the fake class and implements from_metadata with __get_metadata__.
4. We call the from_metadata call and continue tracing.

Reply:
For dynamic shapes, user could call mark_dynamic in FakeFoo.from_real (or from_metadata).